app/models/attachment.rb

```diff
@@ -13,6 +13,49 @@ class Attachment < ActiveRecord::Base
   #   The path to the attachment contents.
   delegate :url, :path, to: :file_upload
 
+  class << self
+    # Calculates the SHA256 hash and return the existing attachment which has the same hash.
+    # Otherwise a new attachment will be built.
+    #
+    # @param [Hash] attributes The hash attributes with the file.
+    # @return [Attachment] The attachment which contains the file.
+    def find_or_initialize_by(attributes, &block)
+      file = attributes[:file]
+      if file
+        digest = file_digest(file)
```

Comment:
I don't think you should change control flow. There are other attributes inside that the user might specify. Instead, you should remove the file key and value, and add the name and file_upload attributes to the attributes, then call super.

```ruby
def find_or_initialize_by(attributes, &block)
  file = attributes.delete(:file)
  attributes[:name] = file_digest(file) if file
  super
end
```

If a file upload is given, maybe we need to override `find_by`?